### ISSUE TITLE
Redesign polymorphic model rendering

### ISSUE THREAD

**Issue by casualjim** [go-swagger/go-swagger#232]
The first question is does it need to be redesigned?

Currently when a type has a discriminator field we turn it into an interface.

This interface defines GetXxx/SetXxx methods.
In subtypes we then generate unexported fields and use the interface methods to satisfy the interface and the unexported field values.

```go
type Pet interface {
   GetName() string
   SetName(string) Pet
   PetType() string
}

type Dog struct {
   nameField string
}

func (d *Dog) GetName() string {
  return d.nameField
}

func (d *Dog) SetName(name string) Pet {
  d.nameField = name
  return d
}

func (d *Dog) PetType() string {
  return "dog"
}
```

This first approach has the disadvantage that there is no struct for the base type, so you need to go through the method calls.

An alternative way of implementing this could be:

```go
type Pet struct {
    Name string
    PetType string
}

func (p *Pet) Pet() *Pet {
  return p
}

type IsPet interface {
    Pet() *Pet
}

type Dog struct {
    *Pet

    OtherAttr string
}
```

In this case the sub type won't be able to stand in for a pet without an explicit method call.

Please give your comments or ideas

**Comment by fredbi** [comment]
**TL;DR**: how about base type implementing MarshalJSON(), MarshalBinary() interfaces?

I've been experimenting lately with polymorphic types, and tried to fix some things.

I've nothing against against the current representation with an interface, but I noticed two itching things:
1. This interface does not extend encoding interfaces.
2. Validation of the base type fields by sub types iterates through the private representation of the base type, thus duplicating code and breaking the principle of the interface.

> (1) is annoying when you want to compose things from a base type (say a tuple of base types, >or whatever...). Currently, codegen breaks when you start making things more complex than putting the base type as a property (and support for arrays is hacky).

> (2) is annoying because it works well for the specific case of base type as a regular property, but cause many breaks when composed in more complex cases.

Regarding (1), I see the Unmarshal{base type} as useful helpers to slurp a http body, but generalizing its use within model classes is tedious (I've tried...).

BTW I added on my branch the Unmarshal{Base type}Map() sort, which was missing.

**Comment by casualjim** [comment]
duplication doesn't matter in generated code it's the same as inlining
This is for types (so only objects nothing else) with a discriminator field, it's not meant to be open ended. In fact it's closed within the domain of a spec. All sub types need to be accounted for.

I don't understand what you mean by 2, can you show an example?

**Comment by fredbi** [comment]
Then I find difficult to twist the model templates toward making a distinction between "IsBaseType" (I am in the schema of the base type"), "IsSubType" (I am a schema extending this base type) and "I am using the base type" (but I don't need everything is transformed as method calls in {{ valueExpression }}. Maybe that was the original intend with "HasBaseType"??
